pkg/scheduler/framework/plugins/interpodaffinity/filtering.go

```diff
@@ -298,6 +298,11 @@ func (pl *InterPodAffinity) RemovePod(ctx context.Context, cycleState *framework
 }
 
 func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
+	// Add read lock so that out-of-tree plugins can mutate their state entry
+	// race-free via cycleState#Lock()/Unlock().
+	cycleState.RLock()
+	defer cycleState.RUnlock()
+
 	c, err := cycleState.Read(preFilterStateKey)
```

Comment:
Why not add the lock in the `Read` function?

Reply:
I was wondering the same, but I think the issue here is that adding the rlock inside Read syncs access to the map only, not the value.

Reply:
but this is not about locking the value.

Reply:
True. Adding RLock to Read but not lock in the Write function is strange.

Reply:
> Why not add the lock in the `Read` function?

An enforced read lock inside `Read` is inappropriate as users should have the option to choose what kind of lock around `Read` - either simply read lock like the in-tree plugins, or write lock as the original issue stated.

Reply:
I would add a function to cycleState with a lock. Like `ReadWithRLock`?

Reply:
> An enforced read lock inside `Read` is inappropriate as users should have the option to choose what kind of lock around `Read`

But there's no choice. We don't need the lock in the in-tree plugins, but we have to have it to avoid breaking out-of-tree plugins.

Reply:
The following will cause a deadlock:

```
cycleState.Lock()
c, err := cycleState.Read(key)
update(c)
cycleState.Write(key, c)
```

Reply:
not that this is something anyone should do...

Reply:
@ahg-g is right, adding an RLock inside would cause double locking - which is an anti-pattern in Go, although I'm not 100% sure it will cause deadlock. Here is what I want to achieve:

- Read pattern
    - explicit read lock around `Read()`
    - read it and process
- Write pattern
    - explicit write lock around `Read()`
    - mutate it during processing

Instead of:

- Write pattern
    - explicit write lock around `Read()`, but there is a read lock under the hood.

Reply:
Wei, the point is that the example I gave shouldn't be deployed at all, the lock should be acquired when accessing the map only, not the values. So what plugins should be doing is:

```
cycleState.RLock()
c, err := cycleState.Read(key)
cycleState.RUnLock()

// use a state specific lock if needed.
c = update(c)

cycleState.Lock()
cycleState.Write(key, c)
cycleState.UnLock()
```

The cycleState lock shouldn't be acquired to synchronize updating the value, it should be acquired only when reading/writing to the map.

Reply:
so locking inside Read and Write makes sense. But I don't want to add a read lock to Read and leave Write as is without locking. So we either adding explicit function that indicates a lock is acquired, or we add locking to both, which I think should also have no material impact on performance for in-tree plugins.

Reply:
Actually, it seems that the place where we potentially have the most number of uses of `Read` is during preemption. Did you check those tests @Huang-Wei ?

If the increase is significant, we can have separate `Read` and `RLockRead` functions. And then we get rid of the `Lock` functions, as there shouldn't be anything using that lock for any other purpose.

Reply:
> the lock should be acquired when accessing the map only, not the values.

Yes, that's true. And that leads to a "best practice" - which is: modifying the map by introducing a "placeholder" key/value pair during PreFilter/PreScore, and then use **plugin-specific lock** in Filter/Score to mutate that **value**.

> Wei, the point is that the example I gave shouldn't be deployed at all,

We cannot guarantee everyone is aware and technically following the pattern. If users directly mutating the **map** during Filter/Score, which is running in parallel, the scheduler would panic. That's why we're introducing read locks in in-tree plugins, isn't it?

> The cycleState lock shouldn't be acquired to synchronize updating the value

It's what [volume_binding#Filter()](https://github.com/kubernetes/kubernetes/blob/0644c7f413f5c5985eec97378477ac4cfec0614a/pkg/scheduler/framework/plugins/volumebinding/volume_binding.go#L222-L224) is doing right now, which adheres to the best practice described above.

But again, what if a plugin doesn't follow this pattern and directly mutating the map?

>  So we either add an explicit function that indicates a lock is acquired (ReadWithRLock), or we add locking to both Read (RLock) and Write (Lock), which I think should have no material impact on performance for in-tree plugins.

Given that PreFilter/PreScore plugins are running in parallel and in-tree plugins only perform Write() in those two phases, I'm reluctant to add locking to Write(), which is pointless and would definitely add overhead.

---

So my point is: if users are adopting the "best practice" above, (basically) all they need is just Read lock during Filter/Score, accompanied with plugin-specific locks. In this case, we even don't need this PR.

But if users don't follow the "best practice", i.e., mutating the map in Filter/Score, what they do is read the map first, and write the value to the map. So lock-free Read() and Write() plus write locks around it would make more sense; instead of (implicit) read lock plus write lock.

> So what plugins should be doing is:
> 
> ```
> cycleState.RLock()
> c, err := cycleState.Read(key)
> cycleState.RUnLock()
> 
> // use a state specific lock if needed.
> c = update(c)
> 
> cycleState.Lock()
> cycleState.Write(key, c)
> cycleState.UnLock()
> ```

If a user has the `c` in place, they don't need to write it back b/c `c` is a pointer already:

```
cycleState.RLock()
c, err := cycleState.Read(key)
cycleState.RUnLock()

// use a state specific lock if needed.
c = update(c)
```

Reply:
OTOH, if we don't have the locks in `Read`, any out-of-tree plugin could break another out-of-tree plugin.

So I would say: we either don't have this PR or we add the lock inside the Read function. I'm also in favor of Abdullah's suggestion of 2 methods: one with locking and one without.

Reply:
> we either don't have this PR or we add the lock inside the Read function.

The "either" part is correct: if everyone follows the "good practice", technically we don't need this PR. The "or" part is "conditionally" correct - when users don't program the plugin using the anti-pattern described above (read and also write to map in a Filter/Score plugin).

> I'm also in favor of Abdullah's suggestion of 2 methods: one with locking and one without.

I'm fine with simplifying this PR by introducing a `SafeRead()` function. `Read()` may not be needed anymore, but it's good to keep it as out-of-tree plugins may be using it. In terms of `SafeWrite()`, it is not needed for now but maybe it's harmless to create it. WDYT?

Reply:
> But again, what if a plugin doesn't follow this pattern and directly mutating the map?

Plugins can't mutate the map itself directly, it is a private member and can only be accessed via Read/Write funcs. Any synchronization needed on the value itself should be done via an external lock specific to that piece of data. Using the CycleState lock makes no sense.

> which is pointless and would definitely add overhead
It is not pointless if we want to provide proper threadsafe access to the map, and it doesn't add overhead if cycleState is properly used: plugins shouldn't be writing a lot to the cycleState map itself, each plugin should do it once per scheduling cycle, and that causes near zero locking overhead. 

The reason I want to have locking done inside both read and write funcs is because I want to remove all Lock related functions, they are confusing and will likely be misused and cause problems.

> If a user has the c in place, they don't need to write it back b/c c is a pointer already:

Sure, the point is that cycleState.Lock is mutex on accessing the map, not the values themselves, which are always references to an opaque data type.

Reply:
@ahg-g I think we're on the same page of how to efficiently use CycleState. What we're talking about here is how to deal with misuse of CycleState, from the perspective of defensive programming.

> because I want to remove all Lock related functions, they are confusing and will likely be misused and cause problems.

This is fair. So @alculquicondor @ahg-g what's your take on this:

- keep 4 functions: Read/SafeRead/Write/SafeWrite, and remove Lock/Unlock
    - in terms of sequential phases (PreFilter/PreScore), use Write or SafeWrite?
- keep 2 functions: Read/Write, and remove Lock/Unlock - i.e., we only have locking version of read/write.

Reply:
I say, let's start with option 2 and evaluate performance. Please make sure preemption performance is not affected as much. I think we would lock for each Pod that is added/removed.

Reply:
sounds good, we also need to Lock inside `Delete` (not sure if anyone uses it).

Reply:
SG, will update and run the perf test.

Reply:
Updated the PR and testing result.

Note that the value of `scheduler_e2e_scheduling` varies a lot due to unpredictable binding duration. In my test, I tried my best to exclude the obvious outliers, but it's still challenging for this metric.